src/WindowStateSaver.vala

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2023 elementary, Inc. <https://elementary.io>
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ */
+
+public class Gala.WindowStateSaver : GLib.Object {
+    private static unowned WindowTracker window_tracker;
+    private static GLib.HashTable<string, GLib.Array<Meta.Window?>> app_windows;
+    private static Sqlite.Database db;
+
+    public static void init (WindowTracker window_tracker) {
+        WindowStateSaver.window_tracker = window_tracker;
+        app_windows = new GLib.HashTable<string, GLib.Array<Meta.Window?>> (GLib.str_hash, GLib.str_equal);
+
+        var path = Path.build_filename (Environment.get_home_dir (), ".local", "share", "io.elementary.gala-windowstate.db");
+        var rc = Sqlite.Database.open_v2 (path, out db);
+
+        if (rc != Sqlite.OK) {
+            critical ("Cannot open database: %d, %s", rc, db.errmsg ());
+            return;
+        }
+
+        Sqlite.Statement stmt;
+        rc = db.prepare_v2 (
+            """
+            CREATE TABLE IF NOT EXISTS apps (
+                id           INTEGER PRIMARY KEY,
+                app_id       TEXT,
+                window_index INTEGER,
```

Comment:
What do you think about using `(app_id, window_index)` as the primary key and always upserting them, to avoid having duplicates in the database?